docs/specs/m2m.md

```diff
@@ -0,0 +1,513 @@
+
+This section includes reviews on prior implementations by competitors.
+
+### Auth0
+
+In Auth0 M2M authentication and authorization, the following concepts must be known first:
+
+- There are API, Permission, Application, User, Role, Organization.
+- An API has 0 or more Permissions.
+- Each Permission is associated with exactly one API.
+- Permission belonging to different API **CAN** share the same name. They ARE NOT considered equal.
+- A M2M Application **MUST** have at least one API.
+- A M2M Application **MAY** have no granted Permissions from any APIs associated with it.
+- A Role can be associated with a Permission of an API **globally**.
+- A User can be associated with a Permission of an API **directly** **globally**.
+- A User can be associated with a Role **organizationally**. In this case, the User inherit the Permissions the Role has **organizationally**.
+- A M2M Application can further has its access limited **organizationally** if you are on a paid plan. See https://auth0.com/docs/manage-users/organizations/organizations-for-m2m-applications/configure-your-application-for-m2m-access#define-organization-behavior
+
+Other facts:
+
+- In Authorization Code Flow, if `audience` is unspecified, the returned `access_token` is **NOT** a JWT. See https://community.auth0.com/t/opaque-versus-jwt-access-token/31028
+- The `permissions` claim is only available if the API has "Enable RBAC" **AND** "Add Permissions in the Access Token".
+- In Authorization Code Flow, Permission is `permissions`.
+- In Client Credentials Flow, Permission is `scope`.
+- When Organization is involved, the globally assigned Roles and Permissions to a User is ignored. `permissions` will always be an empty array. See https://community.auth0.com/t/organization-permissions-claim-empty/99135
+- In Client Credentials Flow, `audience` is required instead of `resource`. And `audience` can only appear once.
+  - This means the JWT access token is always associated with **ONE** `audience`.
+  - Since the JWT access token is always associated with one `audience`, `scope` does not have ambiguity even Permission belonging to different API can share the same name.
+
+An example of the JWT returned by Auth0 when running the Client Credentials Flow
+
+Payload
+```json
+{
+  "iss": "https://dev-fnc259uk.auth0.com/",
+  "sub": "62Amt65MWTRwkoySX65O8JVT735c8UI2@clients",
+  "aud": "myapi",
+  "iat": 1751011899,
+  "exp": 1751098299,
+  "scope": "write:users",
+  "jti": "jsBXsU6pXyB29Bo2RVG3Ki",
+  "client_id": "62Amt65MWTRwkoySX65O8JVT735c8UI2",
+  "permissions": [
+    "write:users"
+  ]
+}
+```
+
+An example of the JWT returned by Auth0 when running the Authorization Code Flow with Organization involved
+
+Payload
+```json
+{
+  "iss": "https://dev-fnc259uk.auth0.com/",
+  "sub": "auth0|683d622f1a6ac540d22d1409",
+  "aud": [
+    "myapi",
+    "https://dev-fnc259uk.auth0.com/userinfo"
+  ],
+  "iat": 1750926072,
+  "exp": 1751012472,
+  "scope": "openid profile email",
```

Comment:
If we use scope as permission, another problem I can think of is in authorization code flow, do we need to write all scopes in the authorization request?

Reply:
Actually, we will not.

I amended the document a lot, and it now includes `sub/scope`. In order for `sub/scope` to appear, `resource` is required. So without `resource`, the Authorization Code Grant behaves exactly the same. No `scope` nor `sub/scope` are modified or added.

If `resource` is additionally provided to the Authorization endpoint, the resulting access token will have 2 `aud`, one for the `resource`, another one for Authgear. `sub/scope` will contain all scope the `sub` (User) has, with respect to `aud`

In Client Credentials Grant, `resource` is required so `sub/scope` is always present, and its value is just equal to `scope`.